js/builtins/weakrefs/cleanup-prevented-with-reference.optional.any.js

```diff
@@ -0,0 +1,56 @@
+// META script=resources/maybe-garbage-collect.js
+// ├──> maybeGarbageCollectAndCleanupAsync
+// └──> resolveGarbageCollection
+/*---
+esid: sec-finalization-registry.prototype.cleanupSome
+info: |
+  FinalizationRegistry.prototype.cleanupSome ( [ callback ] )
+
+  1. Let finalizationRegistry be the this value.
+  2. If Type(finalizationRegistry) is not Object, throw a TypeError exception.
+  3. If finalizationRegistry does not have a [[Cells]] internal slot, throw a TypeError exception.
+  4. If callback is not undefined and IsCallable(callback) is false, throw a TypeError exception.
+  5. Perform ? CleanupFinalizationRegistry(finalizationRegistry, callback).
+  6. Return undefined.
+---*/
+
+let holdingsList = [];
+function cb(holding) {
+  holdingsList.push(holding);
+};
+let finalizationRegistry = new FinalizationRegistry(function() {});
+
+let referenced = {};
+
+function emptyCells() {
+  let target = {};
+  finalizationRegistry.register(target, 'target!');
+  finalizationRegistry.register(referenced, 'referenced');
+
+  let prom = maybeGarbageCollectAndCleanupAsync(target);
+  target = null;
+
+  return prom;
+}
+
+async function execAsync() {
```

Comment:
@jugglinmike I've used this pattern throughout

Reply:
This pattern is a little hard for me to follow because the exception value weaves between being a rejection and a fulfillment, so it's hard to track the way the promise is fulfilled.

```js
(async () => {
  try {
    // test code
  } catch (error) {
    return error;
  }
})().then((errorOrUndefined) => {
  if (errorOrUndefined && errorOrUndefined !== maybeGarbageCollectAndCleanupAsync.NOT_COLLECTED) {
    throw error;
  }
});
```

If we allow the exception to consistently follow the "rejection" route, then we can simplify the call site and (hopefully) improve legibility:

```js
(async () => {
  // test code
})().catch((error) => { // swapping `catch` for `then`
  if (error !== maybeGarbageCollectAndCleanupAsync.NOT_COLLECTED) {
    throw error;
  }
});
```